jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -6609,36 +6613,161 @@ private static char convEndian(boolean isBigEndian, char value) {
+	 * @return value in obj at offset before this operation. This will be v1 if the exchange was successful
+	 */
 	public final <V> Object compareAndExchangeValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		synchronized (inlineTypesLock) {
+			return compareAndExchangeValueNonAtomic(obj, offset, clz, v1, v2);
+		}
 	}
 
+	/**
+	 * Atomically sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of getAcquire.
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if compare is successful
+	 * @return value in obj at offset before this operation. This will be v1 if the exchange was successful
+	 */
 	public final <V> Object compareAndExchangeValueAcquire(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		return compareAndExchangeValue(obj, offset, clz, v1, v2);
 	}
 
+	/**
+	 * Atomically sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of setRelease.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if compare is successful
+	 * @return value in obj at offset before this operation. This will be v1 if the exchange was successful
+	 */
 	public final <V> Object compareAndExchangeValueRelease(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		return compareAndExchangeValue(obj, offset, clz, v1, v2);
 	}
 
+	/**
+	 * Sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if compare is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
 	public final <V> boolean weakCompareAndSetValuePlain(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		return compareAndSetValueNonAtomic(obj, offset, clz, v1, v2);
```

Comment:
weakCompareAndSetValuePlain still as to be atomic. The "CompareAndSetValue" is atomic, the write back to memory is weakly ordered. 

Its functionally correct to use a stronger mode in place of a weaker one, it is not correct to use a weaker mode in place of a stronger one. Basically, all of these are atomic.

Reply:
also, for all of these you only need `compareAndExchangeValueNonAtomic` if it is not flattened

Reply:
> you only need `compareAndExchangeValueNonAtomic` if it is flattened

so in this case I can just use `compareAndSetObject` instead?

Reply:
yes

Reply:
added in 929154f
I created a private helper to call `J9_IS_J9CLASS_FLATTENED` in order to test if `clz` is flattened